File: Week_02/offer40.least-k-numbers.cpp
class Solution {
public:
    vector<int> getLeastNumbers(vector<int>& arr, int k) {
        priority_queue<int> pq;
        vector<int> result(k, 0);
        if (k == 0) {
            return result;
        }
        for (int i = 0; i < k; ++i) {
            pq.push(arr[i]);
        }
        for (int i = k; i < arr.size(); ++i) {
            if (arr[i] < pq.top()) {
                pq.pop();
                pq.push(arr[i]);
            }
        }
        for (int i = k - 1; i >= 0; i--) {
            result[i] = pq.top();
            pq.pop();
        }
        return result;
    }
};

File: Week_02/offer40.least-k-numbers.java
class Solution {
    public int[] getLeastNumbers(int[] arr, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(new Comparator<Integer>() {
            public int compare(Integer t1, Integer t2) {
                return t2 - t1;
            }
        });
        for (int i = 0; i < arr.length; i++) {
            if (pq.size() < k) {
                pq.offer(arr[i]);
            } else if (pq.size() > 0 && pq.size() == k && pq.peek() > arr[i]) {
                pq.poll();
                pq.offer(arr[i]);
            }
        }
        int i = 0;
        int[] result = new int[k];
        while (pq.size() > 0) {
            result[i++] = pq.peek();
            pq.poll();
        }
        return result;
    }
}
